code/modules/logging/log_holder.dm

```diff
@@ -14,51 +13,136 @@ GLOBAL_PROTECT(logger)
 	shutdown = TRUE
 
+/// Iterates over all log category types to assemble them into a tree of main category -> (sub category)[] while also checking for loops and sanity errors
+/datum/log_holder/proc/assemble_log_category_tree()
+	var/static/list/category_tree
+	if(category_tree)
+		return category_tree.Copy()
+
+	category_tree = list()
+	var/list/all_types = subtypesof(/datum/log_category)
+	var/list/known_categories = list()
+	var/list/sub_categories = list()
+
+	// Assemble the master categories
+	for(var/datum/log_category/category_type as anything in all_types)
+		var/category = initial(category_type.category)
+		if(category in known_categories)
+			stack_trace("log category type '[category_type]' has duplicate category '[category]', skipping")
+			continue
+
+		if(!initial(category_type.schema_version))
+			stack_trace("log category type '[category_type]' does not have a valid schema version, skipping")
+			continue
+
+		var/master_category = initial(category_type.master_category)
+		if(master_category)
+			sub_categories[master_category] += list(category_type)
+			continue
+		category_tree[category_type] = list()
+
+	// Sort the sub categories
+	for(var/datum/log_category/master as anything in sub_categories)
+		if(!(master in category_tree))
+			stack_trace("log category [master] is an invalid master category as its a sub category")
+			continue
+		for(var/datum/log_category/sub_category as anything in sub_categories[master])
+			if(initial(sub_category.secret) != initial(master.secret))
+				stack_trace("log category [sub_category] has a secret status that differs from its master category [master]")
+			category_tree[master] += list(sub_category)
+
+	return category_tree.Copy()
+
+/// Initializes the given log category and populates the list of contained categories based on the sub category list
+/datum/log_holder/proc/init_log_category(datum/log_category/category_type, list/datum/log_category/sub_categories)
+	var/datum/log_category/category_instance = new category_type
+
+	var/list/contained_categories = list()
+	for(var/datum/log_category/sub_category as anything in sub_categories)
+		sub_category = new sub_category
+		sub_category.master_category = category_instance
+		log_categories[sub_category.category] = sub_category
+		contained_categories[sub_category.category] = sub_category.schema_version
+
+	log_categories[category_instance.category] = category_instance
+	contained_categories[category_instance.category] = category_instance.schema_version
+
+	var/list/category_header = list(
+		LOG_HEADER_INIT_TIMESTAMP = big_number_to_text(logging_start_timestamp),
+		LOG_HEADER_ROUND_ID = big_number_to_text(GLOB.round_id),
+		LOG_HEADER_SECRET = category_instance.secret,
+		LOG_HEADER_SCHEMA_LIST = contained_categories,
+		LOG_HEADER_CATEGORY = category_instance.category,
+	)
+	rustg_file_write("[json_encode(category_header)]\n", category_instance.get_output_file(null))
+
 /// This is Log because log is a byond internal proc
```

Comment:
This shouldn't be a doc comment btw, this is an implementation detail.

This needed an actual doc comment